ancestrymanager/ancestrymanager.go

```diff
@@ -1,189 +1,263 @@
+	m.errorLogger.Info(fmt.Sprintf("Retrieving ancestry from resource (type=%s)", cai.Type))
+	key := ""
+	orgKey := ""
+	folderKey := ""
 
-		orgID, ok := tfData.GetOk("org_id")
-		if ok && orgID != "" {
-			s := strings.Split(orgID.(string), "/")
-			return append(ancestry, &cloudresourcemanager.Ancestor{
-				ResourceId: &cloudresourcemanager.ResourceId{
-					Type: "organization",
-					Id:   s[len(s)-1],
-				},
-			}), true
+	orgID, orgOK := utils.GetOrganizationFromResource(tfData)
+	if orgOK {
+		orgKey = orgID
+		if !strings.HasPrefix(orgKey, "organizations/") {
+			orgKey = fmt.Sprintf("organizations/%s", orgKey)
+		}
+	}
+	folderID, folderOK := utils.GetFolderFromResource(tfData)
+	if folderOK {
+		folderKey = folderID
+		if !strings.HasPrefix(folderKey, "folders/") {
+			folderKey = fmt.Sprintf("folders/%s", folderKey)
+		}
+	}
+	switch cai.Type {
+	case "cloudresourcemanager.googleapis.com/Folder":
+		if !folderOK {
+			return nil, fmt.Errorf("folder id not found in terraform data")
+		}
+		key = folderKey
+	case "cloudresourcemanager.googleapis.com/Organization":
+		if !orgOK {
+			return nil, fmt.Errorf("organization id not found in terraform data")
+		}
+		key = orgKey
+	case "iam.googleapis.com/Role":
+		// google_organization_iam_custom_role or google_project_iam_custom_role
+		if orgOK {
+			key = orgKey
+		} else {
+			project, err := utils.GetProjectFromResource(tfData, config, *cai, m.errorLogger)
+			if err != nil {
+				return nil, err
+			}
+			key = fmt.Sprintf("projects/%s", project)
+		}
+	case "cloudresourcemanager.googleapis.com/Project", "cloudbilling.googleapis.com/ProjectBillingInfo":
+		projectID, err := utils.GetProjectFromResource(tfData, config, *cai, m.errorLogger)
+		if err != nil {
+			return nil, err
 		}
 
-		folderID, ok := tfData.GetOk("folder_id")
-		if ok && folderID != "" {
-			folderAncestry, err := m.getFolderAncestry(folderID.(string))
+		// Changing project_id forces a new project to be created.
+		// Changing folder_id or org_id forces project to be migrated.
+		// Hence ancestors should not be fetched from API in those scenarios.
+		if orgOK {
+			ancestors := []string{
+				fmt.Sprintf("projects/%s", projectID),
+				fmt.Sprintf("organizations/%s", orgID),
+			}
+			return ancestors, nil
+		}
+		if folderOK {
+			key = folderKey
+			ret, err := m.getAncestorsWithCache(key)
 			if err != nil {
-				m.errorLogger.Error(fmt.Sprintf("Failed to retrieve folder ancestry: %s", err))
-				return nil, false
+				return nil, err
 			}
-			return append(ancestry, folderAncestry...), true
+			ancestors := append([]string{fmt.Sprintf("projects/%s", projectID)}, ret...)
+			return ancestors, nil
 		}
-		return nil, false
+		key = fmt.Sprintf("projects/%s", projectID)
+		ancestors, err := m.getAncestorsWithCache(key)
+		if err != nil {
+			if isGoogleApiErrorWithCode(err, 403) {
+				// new project
+				return []string{key}, nil
```

Comment:
the current behavior for new project seems to be returning 403, convert error. 
example_project_create.tf uses `project_id = "{{.Provider.project}}"`, so I don't think it actually creating a new one.

Reply:
If the field is empty, terraform plan does not pass.
If the field is referencing another variable, eg. random string, then it will be interpreted as empty, and when it detects cai type is project, and retrieving a project id, then it will return error "required field is not set".

If the field is some new value, that returns 403, blocking it to get ancestors or parent should not block it to be converted, right? 

Do you remember which scenario it might cause a side effect? I think we should add such a test case for that.

Reply:
Ah - I see. This is when the user doesn't have permissions to access the project id that's listed.

We should continue to return the 403 in that case, at least for now.

I agree this is behavior we might want to change - maybe as a follow-up to go/tfv-dont-require-projects - but it would be good to talk through what exactly we want the behavior to be (and this PR is already making a lot of other changes).